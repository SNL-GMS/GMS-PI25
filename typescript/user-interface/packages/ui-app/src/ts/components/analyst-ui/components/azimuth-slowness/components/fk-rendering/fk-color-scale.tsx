/* eslint-disable react/destructuring-assignment */
import { FkTypes } from '@gms/common-model';
import { useColorMap } from '@gms/ui-state';
import * as d3 from 'd3';
import React, { useEffect } from 'react';

import { gmsColors } from '~scss-config/color-preferences';

import { createColorScaleImageBitmap } from '../fk-util';

/**
 * FkColorScale Props
 */
export interface FkColorScaleProps {
  minSlow: number;
  maxSlow: number;
  fkUnits: FkTypes.FkUnits;
}

/**
 * The color scale size.
 */
export interface ColorScaleSize {
  width: number;
  height: number;
}

/**
 * FkColorScale Functional Component
 */
export function FkColorScale(props: FkColorScaleProps) {
  const padding = 25;

  /** The color scale size. */
  const colorScaleSize: ColorScaleSize = { width: 80, height: 240 };

  /** Reference to the canvas to draw the color scale. */
  let canvasRef: HTMLCanvasElement | undefined;

  /** The x-axis div container. */
  let xAxisContainerRef: HTMLDivElement | undefined;

  /** Get user preferred color map */
  const [colorMap] = useColorMap();
  /**
   * sets parameters and updates bitmap
   */
  useEffect(() => {
    /**
     * Create and draw the x-axis.
     */
    function createXAxis() {
      if (!xAxisContainerRef) return;
      xAxisContainerRef.innerHTML = '';

      const svg = d3
        .select(xAxisContainerRef)
        .append('svg')
        .attr('width', xAxisContainerRef.clientWidth)
        .attr('height', xAxisContainerRef.clientHeight)
        .style('fill', gmsColors.gmsMain);

      const svgAxis = svg.append('g').attr('class', 'fk-axis');
      const x =
        props.fkUnits === FkTypes.FkUnits.FSTAT
          ? d3
              .scaleLinear()
              .domain([props.minSlow, props.maxSlow])
              .range([padding, xAxisContainerRef.clientWidth - padding])
          : d3
              .scaleLog()
              .domain([props.minSlow, props.maxSlow])
              .range([padding, xAxisContainerRef.clientWidth - padding]);
      const range = props.maxSlow - props.minSlow;
      const tickSize = 7;
      const rangeOfScaleInRealPx = xAxisContainerRef.clientWidth - padding - padding;
      const logarithmicHalfOfScale = rangeOfScaleInRealPx / 2;
      const logarithmicQuarterOfScale = rangeOfScaleInRealPx / 4;
      const logarithmicThreeQuarterOfScale = (rangeOfScaleInRealPx * 3) / 4;

      const xAxis =
        props.fkUnits === FkTypes.FkUnits.FSTAT
          ? d3
              .axisBottom(x)
              .tickSize(tickSize)
              .tickValues([
                props.minSlow,
                props.minSlow + range / 4,
                props.minSlow + range / 2,
                props.minSlow + (range * 3) / 4,
                props.maxSlow
              ])
              .tickFormat(d3.format('.2'))
          : d3
              .axisBottom(x)
              .tickSize(tickSize)
              .tickValues([
                x.invert(padding),
                x.invert(padding + logarithmicQuarterOfScale),
                x.invert(padding + logarithmicHalfOfScale),
                x.invert(padding + logarithmicThreeQuarterOfScale),
                x.invert(xAxisContainerRef.clientWidth - padding)
              ])
              .tickFormat(d3.format('.2'));
      svgAxis.call(xAxis);
    }

    /**
     * Draws the image to the context
     */
    function draw(ctx: CanvasRenderingContext2D, currentImage: ImageBitmap) {
      if (canvasRef) {
        const height = 50;
        canvasRef.width = xAxisContainerRef.clientWidth - padding * 2;
        canvasRef.height = height;
        ctx.drawImage(currentImage, 0, 0, canvasRef.width, height);

        createXAxis();
      }
    }

    async function updateBitmap() {
      /** Canvas rendering context used to draw the color scale. */
      const ctx: CanvasRenderingContext2D = canvasRef.getContext('2d');
      ctx.imageSmoothingEnabled = true;

      /** The current color scale represented as an ImageBitmap. */
      const currentImage: ImageBitmap = await createColorScaleImageBitmap(
        colorScaleSize.width,
        colorScaleSize.height,
        colorMap
      );
      draw(ctx, currentImage);
    }
    if (xAxisContainerRef) {
      // TODO: add a logger
      // eslint-disable-next-line no-console
      updateBitmap().catch(console.error);
    }
  }, [
    canvasRef,
    colorMap,
    colorScaleSize.height,
    colorScaleSize.width,
    props.fkUnits,
    props.maxSlow,
    props.minSlow,
    xAxisContainerRef,
    xAxisContainerRef?.clientWidth
  ]);

  return (
    <div className="fk-color-scale">
      <div
        className="fk-color-scale__xaxis"
        ref={ref => {
          xAxisContainerRef = ref;
        }}
      />
      <canvas
        className="fk-color-scale__canvas"
        ref={ref => {
          canvasRef = ref;
        }}
      />
      {props.fkUnits === FkTypes.FkUnits.POWER ? (
        <div className="fk-color-scale__units">(db)</div>
      ) : null}
    </div>
  );
}
