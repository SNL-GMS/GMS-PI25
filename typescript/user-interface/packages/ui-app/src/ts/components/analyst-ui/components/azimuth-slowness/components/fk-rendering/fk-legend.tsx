import React, { useEffect } from 'react';

import { gmsColors } from '~scss-config/color-preferences';

import * as fkUtil from '../fk-util';

/**
 * Class that renders the FK Display legend which describes peak, predicted, and analyst
 */
export function FkLegend() {
  /** Reference to the canvas to draw different dots for the legend */
  let canvasRef: HTMLCanvasElement | undefined;

  /**
   * Invoked when the component mounted.
   */
  useEffect(() => {
    /**
     * Draws the fk display legend
     */
    function drawFkLegend(ctx: CanvasRenderingContext2D) {
      const xForDots = 6;
      const yForAnalystDot = 12;
      const yForPredictedDot = 28;
      const yForCrossHairDot = 44;
      const canvasHeight = 80;
      const canvasWidth = 20;

      canvasRef.width = canvasWidth;
      canvasRef.height = canvasHeight;

      fkUtil.drawCircle(
        ctx,
        xForDots,
        yForAnalystDot,
        [fkUtil.markerRadiusSize],
        gmsColors.gmsRecessed,
        true
      );
      fkUtil.drawCircle(
        ctx,
        xForDots,
        yForPredictedDot,
        [fkUtil.markerRadiusSize],
        gmsColors.gmsMain,
        true
      );
      fkUtil.drawCircle(
        ctx,
        xForDots,
        yForCrossHairDot,
        [fkUtil.markerRadiusSize],
        gmsColors.gmsMain,
        true
      );
      fkUtil.drawCrosshairDot(
        ctx,
        xForDots,
        yForCrossHairDot,
        gmsColors.gmsRecessed,
        fkUtil.markerRadiusSize
      );
    }

    if (canvasRef) {
      /** Canvas rendering context used to draw different dots for the legend */
      const ctx = canvasRef.getContext('2d');
      drawFkLegend(ctx);
    }
  }, [canvasRef]);

  /**
   * Renders the component.
   */
  return (
    <div className="fk-legend">
      <canvas
        className="fk-legend__canvas"
        ref={ref => {
          canvasRef = ref;
        }}
      />
      <div className="fk-legend-font__analyst">Selected</div>
      <div className="fk-legend-font__peak">Peak</div>
      <div className="fk-legend-font__predicted">Predicted</div>
    </div>
  );
}
