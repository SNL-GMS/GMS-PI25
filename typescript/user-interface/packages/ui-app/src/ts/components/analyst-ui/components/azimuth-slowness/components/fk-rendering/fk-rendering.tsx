/* eslint-disable react/destructuring-assignment */
import { PopoverInteractionKind, Position } from '@blueprintjs/core';
import { Popover2 } from '@blueprintjs/popover2';
import type { ColorTypes, EventTypes } from '@gms/common-model';
import { FkTypes } from '@gms/common-model';
import type { Point } from '@gms/ui-util';
import isEqual from 'lodash/isEqual';
import React from 'react';

import * as fkUtil from '../fk-util';
import { FkColorScale } from './fk-color-scale';
import { FkLegend } from './fk-legend';
/**
 * FkRendering Props
 */
export interface FkRenderingProps {
  signalDetectionFeaturePredictions: EventTypes.FeaturePrediction[];
  analystSelectedPoint: Point; // Scaled selected point
  fkUnitDisplayed: FkTypes.FkUnits;
  renderingHeightPx: number;
  renderingWidthPx: number;
  currentMovieSpectrumIndex: number;
  selectedFk: FkTypes.FkPowerSpectra;
  colorMap: ColorTypes.ColorMapName;
  updateCurrentFk(point: Point): void;
}

/**
 * FkRendering State
 */
export interface FkRenderingState {
  // the min/max value in the fk data
  minFkValue: number;
  maxFkValue: number;
  currentFkDisplayData: number[][];
  selectedPoint: Point; // actual (not scaled x,y position)
}

// The height of the rendering needs to be 8 pixels smaller than the width
export const FK_RENDERING_HEIGHT_OFFSET = 8;

/**
 * FkRendering Component
 */
export class FkRendering extends React.Component<FkRenderingProps, FkRenderingState> {
  /** Reference to the canvas to draw the fk. */
  private canvasRef: HTMLCanvasElement | undefined;

  /** The current fk represented as an ImageBitmap. */
  private currentImage: ImageBitmap | undefined;

  /** The y-axis div container. */
  private yAxisContainerRef: HTMLDivElement | undefined;

  /** The x-axis div container. */
  private xAxisContainerRef: HTMLDivElement | undefined;

  public constructor(props: FkRenderingProps) {
    super(props);
    this.state = {
      minFkValue: 0,
      maxFkValue: 1,
      currentFkDisplayData: [], // Init empty (double array)
      selectedPoint: undefined
    };
  }

  // ***************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ***************************************

  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    return (
      <div
        className="fk"
        style={{
          height: `${this.props.renderingHeightPx + FkTypes.Util.SIZE_OF_FK_RENDERING_AXIS_PX}px`,
          width: `${this.props.renderingWidthPx + FkTypes.Util.SIZE_OF_FK_RENDERING_AXIS_PX}px`
        }}
      >
        <div className="fk-rendering">
          <div
            ref={ref => {
              this.yAxisContainerRef = ref;
            }}
            className="fk-rendering__y-axis"
          />
          <div
            ref={ref => {
              this.xAxisContainerRef = ref;
            }}
            className="fk-rendering__x-axis"
          />
          <div className="fk-rendering__slowness">
            <Popover2
              interactionKind={PopoverInteractionKind.CLICK}
              position={Position.BOTTOM}
              content={
                <div className="fk-rendering__slowness-label">
                  <FkLegend />
                  <FkColorScale
                    minSlow={this.state.minFkValue}
                    maxSlow={this.state.maxFkValue}
                    fkUnits={this.props.fkUnitDisplayed}
                  />
                </div>
              }
            >
              <div className="fk-color-scale__button">slowness (s/Â°)</div>
            </Popover2>
          </div>
          <canvas
            className="fk-rendering__canvas"
            data-cy="primary-fk-rendering"
            width={this.props.renderingWidthPx}
            height={this.props.renderingHeightPx - FK_RENDERING_HEIGHT_OFFSET}
            ref={ref => {
              this.canvasRef = ref;
            }}
            onClick={this.onPrimaryFkClick}
          />
        </div>
      </div>
    );
  }

  /**
   * React component lifecycle
   */
  public async componentDidMount(): Promise<void> {
    const predictedPoint = fkUtil.getPredictedPoint(this.props.signalDetectionFeaturePredictions);
    const fkDisplaySpectrum = this.props.selectedFk?.values[this.props.currentMovieSpectrumIndex];
    const fkDisplayData = fkUtil.getFkHeatmapArrayFromFkSpectra(
      fkDisplaySpectrum,
      this.props.fkUnitDisplayed
    );
    const [min, max] = fkUtil.computeMinMaxFkValues(fkDisplayData);
    this.currentImage = await fkUtil.createFkImageBitmap(
      fkDisplayData,
      min,
      max,
      this.props.colorMap
    );
    // Guard against canvas ref is not yet set (happens in render)
    if (this.canvasRef && this.props.selectedFk) {
      this.drawAllComponents(predictedPoint);
      this.setState({
        maxFkValue: max,
        minFkValue: min,
        currentFkDisplayData: fkDisplayData
      });
    }
  }

  /**
   * React component lifecycle
   */
  public async componentDidUpdate(prevProps: FkRenderingProps): Promise<void> {
    if (this.state.selectedPoint && !this.props.analystSelectedPoint) {
      this.setState({
        selectedPoint: undefined
      });
    }
    if (!this.props.selectedFk) {
      return;
    }
    const predictedPoint = fkUtil.getPredictedPoint(this.props.signalDetectionFeaturePredictions);
    const newFkDisplayData = fkUtil.getFkHeatmapArrayFromFkSpectra(
      this.props.selectedFk.values[this.props.currentMovieSpectrumIndex],
      this.props.fkUnitDisplayed
    );

    if (!isEqual(this.state.currentFkDisplayData, newFkDisplayData)) {
      const [min, max] = fkUtil.computeMinMaxFkValues(newFkDisplayData);
      this.currentImage = await fkUtil.createFkImageBitmap(
        newFkDisplayData,
        min,
        max,
        this.props.colorMap
      );
      this.drawAllComponents(predictedPoint);
      this.setState({
        maxFkValue: max,
        minFkValue: min,
        currentFkDisplayData: newFkDisplayData
      });
    } else if (
      this.props.renderingHeightPx !== prevProps.renderingHeightPx ||
      this.props.renderingWidthPx !== prevProps.renderingWidthPx
    ) {
      this.drawAllComponents(predictedPoint);
    }
  }

  /**
   *
   * @param predictedPoint
   */
  private readonly drawAllComponents = (predictedPoint: FkTypes.AzimuthSlowness) => {
    fkUtil.drawImage(this.canvasRef, this.currentImage);
    fkUtil.draw(
      this.canvasRef,
      this.props.selectedFk,
      predictedPoint,
      this.props.currentMovieSpectrumIndex,
      false,
      false,
      this.state.selectedPoint
    );
    // Draw the x and y axis
    fkUtil.createXAxis(this.props.selectedFk, this.xAxisContainerRef);
    fkUtil.createYAxis(this.props.selectedFk, this.yAxisContainerRef);
  };

  /**
   * When primary fk is clicked, will draw black circle
   */
  private readonly onPrimaryFkClick = async (e: React.MouseEvent<HTMLCanvasElement>) => {
    if (!this.props.selectedFk) {
      return;
    }
    const x = e.clientX - e.currentTarget.getBoundingClientRect().left;
    const y = e.clientY - e.currentTarget.getBoundingClientRect().top;
    const selectedPoint = { x, y };
    const predictedPoint = fkUtil.getPredictedPoint(this.props.signalDetectionFeaturePredictions);

    const { currentFkDisplayData } = this.state;
    const [min, max] = fkUtil.computeMinMaxFkValues(currentFkDisplayData);
    this.currentImage = await fkUtil.createFkImageBitmap(
      currentFkDisplayData,
      min,
      max,
      this.props.colorMap
    );

    if (this.props.updateCurrentFk) {
      // Converting x y point from graphics space to coordinate space
      const scaledXY = fkUtil.convertGraphicsXYtoCoordinate(
        x,
        y,
        this.props.selectedFk,
        this.canvasRef.width,
        this.canvasRef.height
      );
      this.props.updateCurrentFk({ x: scaledXY.x, y: scaledXY.y });
    }
    this.setState(
      {
        selectedPoint
      },
      () => this.drawAllComponents(predictedPoint)
    );
  };
}
