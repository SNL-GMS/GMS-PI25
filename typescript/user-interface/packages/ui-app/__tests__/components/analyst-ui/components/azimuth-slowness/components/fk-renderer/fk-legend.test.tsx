import React from 'react';
import renderer from 'react-test-renderer';

import { FkLegend } from '../../../../../../../src/ts/components/analyst-ui/components/azimuth-slowness/components/fk-rendering/fk-legend';

it('FkLegend renders & matches snapshot', () => {
  const tree = renderer
    .create(
      <div>
        <FkLegend />
      </div>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
