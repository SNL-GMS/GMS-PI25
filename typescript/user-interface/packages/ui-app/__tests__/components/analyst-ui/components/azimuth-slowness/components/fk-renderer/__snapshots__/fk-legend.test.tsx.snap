// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FkLegend renders & matches snapshot 1`] = `
<div>
  <div
    className="fk-legend"
  >
    <canvas
      className="fk-legend__canvas"
    />
    <div
      className="fk-legend-font__analyst"
    >
      Selected
    </div>
    <div
      className="fk-legend-font__peak"
    >
      Peak
    </div>
    <div
      className="fk-legend-font__predicted"
    >
      Predicted
    </div>
  </div>
</div>
`;
